From c87f609d3733f7eb57e5804e8a7cb3810fdd0a22 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 12 Nov 2021 11:14:37 +0800
Subject: [PATCH 19/41] waylandsink: Support NV12_10LE40 NV16_10LE40 and
 NV12|NV12_10LE40|NV16 (AFBC)

Tested on RK356x with:
export GST_MPP_VIDEODEC_DEFAULT_ARM_AFBC=1
gst-play-1.0 video.mp4 --videosink=waylandsink

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 ext/gtk/gstgtkwaylandsink.c                 |  5 +-
 ext/wayland/gstwaylandsink.c                | 97 ++++++++++++++++++++-
 gst-libs/gst/wayland/gstwldisplay.c         |  9 +-
 gst-libs/gst/wayland/gstwldisplay.h         |  2 +-
 gst-libs/gst/wayland/gstwllinuxdmabuf.c     | 46 ++++++++--
 gst-libs/gst/wayland/gstwlvideobufferpool.c |  3 +-
 gst-libs/gst/wayland/gstwlvideoformat.h     |  4 +-
 7 files changed, 145 insertions(+), 21 deletions(-)

diff --git a/ext/gtk/gstgtkwaylandsink.c b/ext/gtk/gstgtkwaylandsink.c
index 48f5004..f749f70 100644
--- a/ext/gtk/gstgtkwaylandsink.c
+++ b/ext/gtk/gstgtkwaylandsink.c
@@ -964,7 +964,7 @@ gst_gtk_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   /* validate the format base on the memory type. */
   if (use_dmabuf) {
     if (!gst_wl_display_check_format_for_dmabuf (priv->display,
-            &priv->drm_info))
+            priv->drm_info.drm_fourcc, priv->drm_info.drm_modifier))
       goto unsupported_drm_format;
   } else if (!gst_wl_display_check_format_for_shm (priv->display,
           &priv->render_info)) {
@@ -1246,7 +1246,8 @@ gst_gtk_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
       "buffer %" GST_PTR_FORMAT " does not have a wl_buffer from our "
       "display, creating it", buffer);
 
-  if (gst_wl_display_check_format_for_dmabuf (priv->display, &priv->drm_info)) {
+  if (gst_wl_display_check_format_for_dmabuf (priv->display,
+      priv->drm_info.drm_fourcc, priv->drm_info.drm_modifier)) {
     guint i, nb_dmabuf = 0;
 
     for (i = 0; i < gst_buffer_n_memory (buffer); i++)
diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 1f093bc..d99e3cb 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -42,9 +42,10 @@
 #include <config.h>
 #endif
 
+#include <drm_fourcc.h>
+
 #include "gstwaylandsink.h"
 
-#include <drm_fourcc.h>
 #include <gst/allocators/allocators.h>
 #include <gst/video/gstvideodmabufpool.h>
 #include <gst/video/videooverlay.h>
@@ -722,6 +723,64 @@ gst_wayland_sink_event (GstBaseSink * bsink, GstEvent * event)
   return ret;
 }
 
+static GstCaps *
+gst_wayland_sink_fixup_caps (GstWaylandSink * self, GstCaps * caps)
+{
+  GstCaps *tmp_caps = NULL;
+
+  if (gst_wl_display_check_format_for_dmabuf (self->display,
+          DRM_FORMAT_NV15, DRM_FORMAT_MOD_LINEAR)) {
+    tmp_caps = gst_caps_from_string (
+        GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_DMABUF,
+            "NV12_10LE40"));
+
+    if (gst_wl_display_check_format_for_dmabuf (self->display,
+            DRM_FORMAT_NV15, DRM_AFBC_MODIFIER)) {
+      gst_caps_ref (tmp_caps);
+      gst_caps_append (caps, tmp_caps);
+
+      gst_caps_set_simple (tmp_caps, "arm-afbc", G_TYPE_INT, 1, NULL);
+    }
+
+    gst_caps_append (caps, tmp_caps);
+  }
+
+  if (gst_wl_display_check_format_for_dmabuf (self->display,
+          DRM_FORMAT_NV20, DRM_FORMAT_MOD_LINEAR)) {
+    tmp_caps = gst_caps_from_string (GST_VIDEO_CAPS_MAKE ("NV16_10LE40"));
+
+    if (gst_wl_display_check_format_for_dmabuf (self->display,
+            DRM_FORMAT_NV20, DRM_AFBC_MODIFIER)) {
+      gst_caps_ref (tmp_caps);
+      gst_caps_append (caps, tmp_caps);
+
+      gst_caps_set_simple (tmp_caps, "arm-afbc", G_TYPE_INT, 1, NULL);
+    }
+
+    gst_caps_append (caps, tmp_caps);
+  }
+
+  if (gst_wl_display_check_format_for_dmabuf (self->display,
+          DRM_FORMAT_NV12, DRM_AFBC_MODIFIER)) {
+    tmp_caps = gst_caps_from_string (
+        GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_DMABUF,
+            "NV12"));
+    gst_caps_set_simple (tmp_caps, "arm-afbc", G_TYPE_INT, 1, NULL);
+    gst_caps_append (caps, tmp_caps);
+  }
+
+  if (gst_wl_display_check_format_for_dmabuf (self->display,
+        DRM_FORMAT_NV16, DRM_AFBC_MODIFIER)) {
+    tmp_caps = gst_caps_from_string (
+        GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_DMABUF,
+            "NV16"));
+    gst_caps_set_simple (tmp_caps, "arm-afbc", G_TYPE_INT, 1, NULL);
+    gst_caps_append (caps, tmp_caps);
+  }
+
+  return caps;
+}
+
 static GstCaps *
 gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
 {
@@ -750,6 +809,8 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     gst_structure_take_value (gst_caps_get_structure (caps, 1), "format",
         &format_list);
 
+    caps = gst_wayland_sink_fixup_caps (self, caps);
+
     GST_DEBUG_OBJECT (self, "display caps: %" GST_PTR_FORMAT, caps);
   }
 
@@ -855,6 +916,8 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
 {
   GstWaylandSink *self = GST_WAYLAND_SINK (bsink);;
   gboolean use_dmabuf;
+  GstStructure *s;
+  gint value;
 
   GST_INFO_OBJECT (self, "set caps %" GST_PTR_FORMAT, caps);
 
@@ -875,12 +938,22 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
       gst_video_info_dma_drm_init (&self->drm_info);
   }
 
+  /* parse AFBC from caps */
+  s = gst_caps_get_structure (caps, 0);
+  if (gst_structure_get_int (s, "arm-afbc", &value)) {
+    if (value)
+      GST_VIDEO_INFO_SET_AFBC (&self->video_info);
+    else
+      GST_VIDEO_INFO_UNSET_AFBC (&self->video_info);
+  }
+
   self->have_mastering_info =
       gst_video_mastering_display_info_from_caps (&self->minfo, caps);
   self->have_light_info =
       gst_video_content_light_level_from_caps (&self->linfo, caps);
 
   self->render_info_changed = TRUE;
+
   self->skip_dumb_buffer_copy = FALSE;
 
   /* free pooled buffer used with previous caps */
@@ -895,7 +968,7 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   /* validate the format base on the memory type. */
   if (use_dmabuf) {
     if (!gst_wl_display_check_format_for_dmabuf (self->display,
-            &self->drm_info))
+            self->drm_info.drm_fourcc, self->drm_info.drm_modifier))
       goto unsupported_drm_format;
   } else if (!gst_wl_display_check_format_for_shm (self->display,
           &self->render_info)) {
@@ -942,12 +1015,20 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
   GstBufferPool *pool = NULL;
   gboolean need_pool;
   guint size;
+  GstStructure *s;
+  gint value;
 
   gst_query_parse_allocation (query, &caps, &need_pool);
 
   if (caps == NULL)
     return FALSE;
 
+  s = gst_caps_get_structure (caps, 0);
+  if (gst_structure_get_int (s, "arm-afbc", &value) && value) {
+    GST_DEBUG_OBJECT (bsink, "no allocation for AFBC");
+    return FALSE;
+  }
+
   if (gst_video_is_dma_drm_caps (caps)) {
     GstVideoInfoDmaDrm drm_info;
 
@@ -1169,7 +1250,8 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
       "buffer %" GST_PTR_FORMAT " does not have a wl_buffer from our "
       "display, creating it", buffer);
 
-  if (gst_wl_display_check_format_for_dmabuf (self->display, &self->drm_info)) {
+  if (gst_wl_display_check_format_for_dmabuf (self->display,
+          self->drm_info.drm_fourcc, self->drm_info.drm_modifier)) {
     guint i, nb_dmabuf = 0;
 
     for (i = 0; i < gst_buffer_n_memory (buffer); i++)
@@ -1222,6 +1304,9 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
   }
 
 handle_shm:
+  if (!wbuf && GST_VIDEO_INFO_IS_AFBC (&self->video_info))
+    goto no_afbc;
+
   if (!wbuf && gst_wl_display_check_format_for_shm (self->display,
           &self->render_info)) {
     if (gst_buffer_n_memory (buffer) == 1 && gst_is_fd_memory (mem))
@@ -1324,6 +1409,12 @@ no_wl_buffer:
     ret = GST_FLOW_ERROR;
     goto done;
   }
+no_afbc:
+  {
+    GST_ERROR_OBJECT (self, "could not import AFBC");
+    ret = GST_FLOW_ERROR;
+    goto done;
+  }
 activate_failed:
   {
     GST_ELEMENT_ERROR (self, RESOURCE, FAILED,
diff --git a/gst-libs/gst/wayland/gstwldisplay.c b/gst-libs/gst/wayland/gstwldisplay.c
index c82ec96..6466a8a 100644
--- a/gst-libs/gst/wayland/gstwldisplay.c
+++ b/gst-libs/gst/wayland/gstwldisplay.c
@@ -22,6 +22,8 @@
 #include <config.h>
 #endif
 
+#include <drm_fourcc.h>
+
 #include "gstwldisplay.h"
 #include "gstwloutput-private.h"
 
@@ -34,7 +36,6 @@
 #include "xdg-shell-client-protocol.h"
 
 #include <errno.h>
-#include <drm_fourcc.h>
 
 #define GST_CAT_DEFAULT gst_wl_display_debug
 GST_DEBUG_CATEGORY_STATIC (GST_CAT_DEFAULT);
@@ -262,7 +263,7 @@ dmabuf_modifier (void *data, struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf,
   if (format == DRM_FORMAT_INVALID || modifier == DRM_FORMAT_MOD_INVALID)
     return;
 #else
-    if (fourcc == DRM_FORMAT_INVALID)
+    if (format == DRM_FORMAT_INVALID)
       return;
 
     if (modifier == DRM_FORMAT_MOD_INVALID)
@@ -504,11 +505,9 @@ gst_wl_display_check_format_for_shm (GstWlDisplay * self,
 
 gboolean
 gst_wl_display_check_format_for_dmabuf (GstWlDisplay * self,
-    const GstVideoInfoDmaDrm * drm_info)
+    const guint fourcc, const guint64 modifier)
 {
   GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
-  guint64 modifier = drm_info->drm_modifier;
-  guint fourcc = drm_info->drm_fourcc;
   GArray *formats, *modifiers;
   guint i;
 
diff --git a/gst-libs/gst/wayland/gstwldisplay.h b/gst-libs/gst/wayland/gstwldisplay.h
index ca6e6a9..a39587a 100644
--- a/gst-libs/gst/wayland/gstwldisplay.h
+++ b/gst-libs/gst/wayland/gstwldisplay.h
@@ -71,7 +71,7 @@ gboolean gst_wl_display_check_format_for_shm (GstWlDisplay * self,
 
 GST_WL_API
 gboolean gst_wl_display_check_format_for_dmabuf (GstWlDisplay * self,
-    const GstVideoInfoDmaDrm *drm_info);
+    const guint fourcc, const guint64 modifier);
 
 GST_WL_API
 struct wl_display *gst_wl_display_get_display (GstWlDisplay * self);
diff --git a/gst-libs/gst/wayland/gstwllinuxdmabuf.c b/gst-libs/gst/wayland/gstwllinuxdmabuf.c
index b1e0163..468c7ae 100644
--- a/gst-libs/gst/wayland/gstwllinuxdmabuf.c
+++ b/gst-libs/gst/wayland/gstwllinuxdmabuf.c
@@ -24,8 +24,12 @@
 #include <config.h>
 #endif
 
+#include <drm_fourcc.h>
+
 #include "gstwllinuxdmabuf.h"
 
+#include <gst/video/video-info-dma.h>
+
 #include "linux-dmabuf-unstable-v1-client-protocol.h"
 
 GST_DEBUG_CATEGORY (gst_wl_dmabuf_debug);
@@ -76,16 +80,27 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
   const gint *strides = NULL;
   GstVideoMeta *vmeta;
   guint nplanes = 0, flags = 0;
+  gfloat stride_scale = 1.0f;
   struct zwp_linux_buffer_params_v1 *params;
   ConstructBufferData data = { 0 };
 
   g_return_val_if_fail (gst_wl_display_check_format_for_dmabuf (display,
-          drm_info), NULL);
+          drm_info->drm_fourcc, drm_info->drm_modifier), NULL);
 
   mem = gst_buffer_peek_memory (buf, 0);
   fourcc = drm_info->drm_fourcc;
   modifier = drm_info->drm_modifier;
 
+  if (!gst_video_info_dma_drm_to_video_info (drm_info, &info)) {
+    GST_ERROR_OBJECT (display, "GstVideoMeta is needed to carry DMABuf using "
+        "'memory:DMABuf' caps feature.");
+    data.wbuf = NULL;
+    goto out;
+  }
+
+  if (GST_VIDEO_INFO_IS_AFBC (&info))
+    modifier = DRM_AFBC_MODIFIER;
+
   vmeta = gst_buffer_get_video_meta (buf);
   if (vmeta) {
     finfo = drm_info->vinfo.finfo;
@@ -94,18 +109,13 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
     nplanes = vmeta->n_planes;
     offsets = vmeta->offset;
     strides = vmeta->stride;
-  } else if (gst_video_info_dma_drm_to_video_info (drm_info, &info)) {
+  } else {
     finfo = info.finfo;
     nplanes = GST_VIDEO_INFO_N_PLANES (&info);
     width = info.width;
     height = info.height;
     offsets = info.offset;
     strides = info.stride;
-  } else {
-    GST_ERROR_OBJECT (display, "GstVideoMeta is needed to carry DMABuf using "
-        "'memory:DMABuf' caps feature.");
-    data.wbuf = NULL;
-    goto out;
   }
 
   GST_DEBUG_OBJECT (display,
@@ -113,6 +123,27 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
       " (%d x %d), DRM fourcc %" GST_FOURCC_FORMAT, gst_buffer_get_size (buf),
       width, height, GST_FOURCC_ARGS (fourcc));
 
+  if (GST_VIDEO_INFO_IS_AFBC (&info)) {
+    /* Mali uses these formats instead */
+    if (fourcc == DRM_FORMAT_NV12) {
+      fourcc = DRM_FORMAT_YUV420_8BIT;
+      nplanes = 1;
+      stride_scale = 1.5;
+    } else if (fourcc == DRM_FORMAT_NV15) {
+      fourcc = DRM_FORMAT_YUV420_10BIT;
+      nplanes = 1;
+      stride_scale = 1.5;
+    } else if (fourcc == DRM_FORMAT_NV16) {
+      fourcc = DRM_FORMAT_YUYV;
+      nplanes = 1;
+      stride_scale = 2;
+    } else {
+      GST_ERROR_OBJECT (mem->allocator, "unsupported format for AFBC");
+      data.wbuf = NULL;
+      goto out;
+    }
+  }
+
   /* Creation and configuration of planes  */
   params = zwp_linux_dmabuf_v1_create_params (gst_wl_display_get_dmabuf_v1
       (display));
@@ -123,6 +154,7 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
 
     offset = offsets[i];
     stride = get_drm_stride (finfo, strides, i);
+    stride *= stride_scale;
     if (gst_buffer_find_memory (buf, offset, 1, &mem_idx, &length, &skip)) {
       GstMemory *m = gst_buffer_peek_memory (buf, mem_idx);
       gint fd = gst_dmabuf_memory_get_fd (m);
diff --git a/gst-libs/gst/wayland/gstwlvideobufferpool.c b/gst-libs/gst/wayland/gstwlvideobufferpool.c
index 3292056..5013190 100644
--- a/gst-libs/gst/wayland/gstwlvideobufferpool.c
+++ b/gst-libs/gst/wayland/gstwlvideobufferpool.c
@@ -22,10 +22,11 @@
 #include <config.h>
 #endif
 
+#include <drm_fourcc.h>
+
 #include "gstwlvideobufferpool.h"
 
 #include <gst/allocators/allocators.h>
-#include <drm_fourcc.h>
 
 GST_DEBUG_CATEGORY (gst_wl_videobufferpool_debug);
 #define GST_CAT_DEFAULT gst_wl_videobufferpool_debug
diff --git a/gst-libs/gst/wayland/gstwlvideoformat.h b/gst-libs/gst/wayland/gstwlvideoformat.h
index 53b579f..5282591 100644
--- a/gst-libs/gst/wayland/gstwlvideoformat.h
+++ b/gst-libs/gst/wayland/gstwlvideoformat.h
@@ -38,12 +38,12 @@ G_BEGIN_DECLS
  */
 #if G_BYTE_ORDER == G_BIG_ENDIAN
 #define GST_WL_VIDEO_FORMATS "{ BGR10A2_LE, RGB10A2_LE, AYUV, RGBA, ARGB, " \
-    "BGRA, ABGR, BGR10x2_LE, RGB10x2_LE, P010_10LE, NV12_10LE40, Y444, v308, " \
+    "BGRA, ABGR, BGR10x2_LE, RGB10x2_LE, P010_10LE, NV12_10LE40, NV16_10LE40, Y444, v308, " \
     "RGBx, xRGB, BGRx, xBGR, RGB, BGR, Y42B, NV16, NV61, YUY2, YVYU, UYVY, " \
     "I420, YV12, NV12, NV21, Y41B, YUV9, YVU9, BGR16, RGB16 }"
 #elif G_BYTE_ORDER == G_LITTLE_ENDIAN
 #define GST_WL_VIDEO_FORMATS "{ BGR10A2_LE, RGB10A2_LE, AYUV, RGBA, ARGB, " \
-    "BGRA, ABGR, BGR10x2_LE, RGB10x2_LE, P010_10LE, NV12_10LE40, Y444, v308, " \
+    "BGRA, ABGR, BGR10x2_LE, RGB10x2_LE, P010_10LE, NV12_10LE40, NV16_10LE40, Y444, v308, " \
     "RGBx, xRGB, BGRx, xBGR, RGB, BGR, Y42B, NV16, NV61, YUY2, YVYU, UYVY, " \
     "I420, YV12, NV12, NV21, Y41B, YUV9, YVU9, BGR16, RGB16 }"
 #endif
-- 
2.39.5

